File: frontend/src/lib/run-progress.ts
import type { Schemas } from '@/api/client'

export type RunResponse = Schemas['RunResponse']
export type PhaseProgress = Schemas['PhaseProgressResponse']

// Stesso ordine della pipeline (app/pipeline.py, app/run_progress.py PHASES).
export const PHASE_ORDER = ['scanning', 'resolving', 'indexing', 'matching', 'executing', 'reconciling'] as const
export type PhaseName = (typeof PHASE_ORDER)[number]

export type StepState = 'done' | 'running' | 'pending' | 'skipped'

export interface Step {
  phase: PhaseName
  state: StepState
  progress: PhaseProgress | undefined
}

// Una fase assente da run.phases non è ancora partita; a run finita (per
// un errore che ha interrotto la pipeline) resta "skipped", mai "pending"
// per sempre.
export function runSteps(run: RunResponse): Step[] {
  const finished = run.finished_at != null
  return PHASE_ORDER.map((phase) => {
    const progress = run.phases?.[phase]
    let state: StepState
    if (progress?.status === 'done') state = 'done'
    else if (progress?.status === 'running' && !finished) state = 'running'
    else if (progress) state = 'done'
    else state = finished ? 'skipped' : 'pending'
    return { phase, state, progress }
  })
}

export function percent(done: number | null | undefined, total: number | null | undefined): number | null {
  if (total == null || total <= 0 || done == null) return null
  return Math.min(100, Math.round((done / total) * 100))
}

// Stima del tempo rimanente della fase in corso dal ritmo osservato finora.
// Nessuna stima nei primi secondi o con pochi elementi: sarebbe solo rumore.
export function etaSeconds(progress: PhaseProgress | undefined, now: number = Date.now()): number | null {
  if (!progress?.started_at || progress.total == null || progress.done <= 0) return null
  const elapsed = (now - new Date(progress.started_at).getTime()) / 1000
  if (elapsed < 5 || progress.done < 3) return null
  const remaining = progress.total - progress.done
  if (remaining <= 0) return 0
  return Math.round((elapsed / progress.done) * remaining)
}

export function formatDuration(seconds: number): string {
  const s = Math.max(0, Math.round(seconds))
  const h = Math.floor(s / 3600)
  const m = Math.floor((s % 3600) / 60)
  const sec = s % 60
  if (h > 0) return `${h}:${String(m).padStart(2, '0')}:${String(sec).padStart(2, '0')}`
  return `${String(m).padStart(2, '0')}:${String(sec).padStart(2, '0')}`
}

export function formatCount(n: number): string {
  return n.toLocaleString('en-US')
}

File: app/run_progress.py
"""Avanzamento live di una run, per il popup di stato (RunStatusIndicator):
per ogni fase quanti elementi sono stati fatti su quanti, un testo di
dettaglio (disco, client o tracker corrente, attesa per un rate limit) e
quando è iniziata — da cui il frontend ricava barra, x/y e stima del tempo
rimanente.

Tutto vive su run_log: `phase_total`/`phase_done`/`phase_detail` per la
fase in corso, `phases_json` per il riepilogo di tutte le fasi (anche
quelle già concluse, mostrate nello stepper). Scritto nel DB al massimo
ogni COMMIT_EVERY elementi o COMMIT_INTERVAL_SECONDS, mai a ogni file: il
poller legge GET /api/runs ogni pochi secondi, un commit per file
rallenterebbe lo scan senza nessun beneficio visibile.
"""

import json
import time
from datetime import UTC, datetime

from sqlalchemy.orm import Session

from app.models import RunLog

PHASES = ("scanning", "resolving", "indexing", "matching", "executing", "reconciling")

COMMIT_EVERY = 50
COMMIT_INTERVAL_SECONDS = 1.0


class RunProgress:
    def __init__(self, session: Session, run: RunLog):
        self._session = session
        self._run = run
        self._phases: dict[str, dict] = json.loads(run.phases_json) if run.phases_json else {}
        self._current: str | None = None
        self._pending = 0
        self._last_commit = 0.0

    # --- fasi -----------------------------------------------------------

    def start_phase(self, phase: str, total: int | None = None, detail: str | None = None) -> None:
        self._close_current()
        self._current = phase
        self._phases[phase] = {
            "status": "running", "done": 0, "total": total, "skipped": 0,
            "started_at": datetime.now(UTC).isoformat(),
        }
        self._run.current_phase = phase
        self._run.phase_total = total
        self._run.phase_done = 0
        self._run.phase_detail = detail
        self._flush()

    def add_total(self, n: int) -> None:
        """Il totale di una fase che si scopre a pezzi (un disco o un
        tracker alla volta)."""
        state = self._state()
        state["total"] = (state.get("total") or 0) + n
        self._run.phase_total = state["total"]
        self._flush()

    def advance(self, n: int = 1, *, skipped: int = 0) -> None:
        state = self._state()
        state["done"] += n
        state["skipped"] += skipped
        self._run.phase_done = state["done"]
        self._pending += n + skipped
        now = time.monotonic()
        if self._pending >= COMMIT_EVERY or now - self._last_commit >= COMMIT_INTERVAL_SECONDS:
            self._flush()

    def detail(self, text: str | None) -> None:
        """Sempre scritto subito: è ciò che l'utente legge per capire
        perché la run sembra ferma (es. attesa per un rate limit)."""
        self._run.phase_detail = text
        self._flush()

    def result(self, *, candidates: int = 0, executed: int = 0) -> None:
        self._run.matches_found = (self._run.matches_found or 0) + candidates
        self._run.auto_executed = (self._run.auto_executed or 0) + executed

    def finish(self) -> None:
        self._close_current()
        self._run.current_phase = None
        self._run.phase_total = None
        self._run.phase_done = None
        self._run.phase_detail = None
        self._flush()

    # --- interni --------------------------------------------------------

    def _state(self) -> dict:
        if self._current is None:
            raise RuntimeError("Nessuna fase in corso")
        return self._phases[self._current]

    def _close_current(self) -> None:
        if self._current is None:
            return
        state = self._phases[self._current]
        state["status"] = "done"
        state["finished_at"] = datetime.now(UTC).isoformat()
        if state.get("total") is None:
            state["total"] = state["done"]

    def _flush(self) -> None:
        self._run.phases_json = json.dumps(self._phases)
        self._session.commit()
        self._pending = 0
        self._last_commit = time.monotonic()


class NullProgress:
    """Stessa interfaccia di RunProgress, nessun effetto: il default dei
    moduli che riportano avanzamento (risoluzione, matching, esecuzione)
    quando vengono chiamati fuori da una run (test, API)."""

    def add_total(self, n: int) -> None:
        pass

    def advance(self, n: int = 1, *, skipped: int = 0) -> None:
        pass

    def detail(self, text: str | None) -> None:
        pass

    def result(self, *, candidates: int = 0, executed: int = 0) -> None:
        pass


NULL_PROGRESS = NullProgress()
